File: deck/seed/code/native/swift/file/runtime/aio.swift
// Asynchronous filesystem runtime for the swift target, over NIOFileSystem -- swift-nio's asynchronous filesystem
// API, which is the ecosystem's answer and not a thread pool of our own.
//
// WHY NOT FOUNDATION. FileManager and FileHandle are synchronous to the bone. Wrapping them in `Task.detached`
// gives an `async` signature over a blocked thread, which reads as asynchronous and is not: a thousand concurrent
// reads become a thousand stalled threads. NIOFileSystem runs the syscalls on a bounded pool with a real async
// interface, the same shape `tokio::fs` has on the rust target, so the two backends behave alike under load.
//
// TOTAL, LIKE THE REST OF THE NATIVE SURFACE. Every call here `throws` upstream and the public file API has no
// error channel, so a failure reads as the empty / false / zero answer. A handle open is the exception: a missing
// file there cannot be papered over.
//
// Reached only through the public file API. See note/term/stdlib/native-async-file-and-server.md.
import Foundation
import NIOCore
import _NIOFileSystem

enum aio {
  // An open file. NIOFileSystem addresses reads and writes by absolute offset and keeps no cursor, so the cursor
  // `seek` moves lives here. A class, not a struct, because the emitted `file` record holds it and every emitted
  // record is a value: two copies of the record must be two views of ONE open file, not two cursors.
  final class Handle: @unchecked Sendable {
    let file: ReadWriteFileHandle
    var at: Int64
    init(file: ReadWriteFileHandle, at: Int64) {
      self.file = file
      self.at = at
    }
  }

  // A read stream: the open file, where it is, and how much of the window is left (nil for "to the end").
  final class Reader: @unchecked Sendable {
    let file: ReadFileHandle
    var at: Int64
    var left: Int64?
    init(file: ReadFileHandle, at: Int64, left: Int64?) {
      self.file = file
      self.at = at
      self.left = left
    }
  }

  final class Writer: @unchecked Sendable {
    let file: WriteFileHandle
    var at: Int64
    init(file: WriteFileHandle, at: Int64) {
      self.file = file
      self.at = at
    }
  }

  private static var shared: FileSystem { FileSystem.shared }

  // ---- whole file ----

  static func fileRead(_ path: String) async -> String {
    guard let bytes = await readBuffer(path) else { return "" }
    return String(buffer: bytes)
  }

  static func fileReadBytes(_ path: String) async -> Data {
    guard let bytes = await readBuffer(path) else { return Data() }
    return Data(bytes.readableBytesView)
  }

  private static func readBuffer(_ path: String) async -> ByteBuffer? {
    do {
      return try await shared.withFileHandle(forReadingAt: FilePath(path)) { file in
        try await file.readToEnd(maximumSizeAllowed: .gibibytes(1))
      }
    } catch {
      return nil
    }
  }

  static func fileWrite(_ path: String, _ data: String) async {
    await writeBytes(path, Array(data.utf8), append: false)
  }

  static func fileWriteBytes(_ path: String, _ data: Data) async {
    await writeBytes(path, Array(data), append: false)
  }

  static func fileAppend(_ path: String, _ data: String) async {
    await writeBytes(path, Array(data.utf8), append: true)
  }

  private static func writeBytes(
    _ path: String,
    _ bytes: [UInt8],
    append: Bool
  ) async {
    do {
      let at = FilePath(path)
      let start: Int64 =
        append ? Int64((try? await shared.info(forFileAt: at)?.size) ?? 0) : 0
      let options: OpenOptions.Write =
        append
        ? .modifyFile(createIfNecessary: true)
        : .newFile(replaceExisting: true)

      try await shared.withFileHandle(forWritingAt: at, options: options) { file in
        _ = try await file.write(
          contentsOf: bytes,
          toAbsoluteOffset: start
        )
      }
    } catch {
      return
    }
  }

  static func fileCopy(_ from: String, _ to: String, _ deep: Bool) async {
    // NIOFileSystem's copyItem is already recursive for a directory, so `deep` only decides whether a directory
    // is allowed through at all: without it this is the single-file copy node's copyFile means.
    do {
      if !deep {
        let info = try await shared.info(forFileAt: FilePath(from))

        if info?.type == .directory {
          return
        }
      }

      try await shared.copyItem(at: FilePath(from), to: FilePath(to))
    } catch {
      return
    }
  }

  static func fileMove(_ from: String, _ to: String) async {
    try? await shared.moveItem(at: FilePath(from), to: FilePath(to))
  }

  static func fileRemove(_ path: String, _ deep: Bool) async {
    _ = try? await shared.removeItem(at: FilePath(path), recursively: deep)
  }

  // an empty `kind` asks only whether the path exists; "file", "directory" and "link" ask what it is
  static func fileTest(_ path: String, _ kind: String) async -> Bool {
    let at = FilePath(path)

    do {
      if kind == "link" {
        let info = try await shared.info(forFileAt: at, infoAboutSymbolicLink: true)
        return info?.type == .symlink
      }

      guard let info = try await shared.info(forFileAt: at) else { return false }

      switch kind {
      case "": return true
      case "file": return info.type == .regular
      case "directory": return info.type == .directory
      default: return true
      }
    } catch {
      return false
    }
  }

  // ---- link ----

  static func linkMake(_ from: String, _ to: String, _ hard: Bool) async {
    // NIOFileSystem has no hard link, and a hard link is one syscall: this is the same `link(2)` any crate calls.
    if hard {
      _ = from.withCString { source in
        to.withCString { target in link(source, target) }
      }

      return
    }

    try? await shared.createSymbolicLink(
      at: FilePath(to),
      withDestination: FilePath(from)
    )
  }

  static func linkRead(_ path: String) async -> String {
    guard
      let target = try? await shared.destinationOfSymbolicLink(at: FilePath(path))
    else {
      return ""
    }

    return target.string
  }

  // ---- permission and owner ----

  static func permissionRead(_ path: String) async -> Int {
    guard let info = try? await shared.info(forFileAt: FilePath(path)) else {
      return 0
    }

    return Int(info.permissions.rawValue)
  }

  static func permissionWrite(_ path: String, _ mode: Int) async {
    _ = path.withCString { chmod($0, mode_t(mode)) }
  }

  static func ownerUser(_ path: String) async -> Int {
    guard let info = try? await shared.info(forFileAt: FilePath(path)) else {
      return 0
    }

    return Int(info.userID.rawValue)
  }

  static func ownerGroup(_ path: String) async -> Int {
    guard let info = try? await shared.info(forFileAt: FilePath(path)) else {
      return 0
    }

    return Int(info.groupID.rawValue)
  }

  static func ownerWrite(_ path: String, _ user: Int, _ group: Int) async {
    _ = path.withCString { chown($0, uid_t(user), gid_t(group)) }
  }

  // ---- temporary ----

  // A unique name under the system temporary directory. The name is a uuid, so there is no retry loop and no
  // mkdtemp template: a collision cannot happen.
  static func temporaryMake(
    _ kind: String,
    _ prefix: String,
    _ suffix: String
  ) async -> String {
    let root =
      (try? await shared.temporaryDirectory)
      ?? FilePath(NSTemporaryDirectory())
    let name = "\(prefix)\(UUID().uuidString.lowercased())\(suffix)"
    let at = root.appending(name)

    if kind == "directory" {
      try? await shared.createDirectory(at: at, withIntermediateDirectories: true)
    } else {
      await writeBytes(at.string, [], append: false)
    }

    return at.string
  }

  // ---- handle ----

  static func handleOpen(
    _ path: String,
    _ read: Bool,
    _ write: Bool,
    _ create: Bool,
    _ append: Bool,
    _ clear: Bool
  ) async -> Handle {
    let at = FilePath(path)
    let options: OpenOptions.Write =
      clear && !append
      ? .newFile(replaceExisting: true)
      : .modifyFile(createIfNecessary: create)

    guard
      let file = try? await shared.openFile(
        forReadingAndWritingAt: at,
        options: options
      )
    else {
      fatalError("file handle open: \(path)")
    }

    // appending starts at the end, everything else at the start, which is what the open flags mean elsewhere
    let start: Int64 =
      append ? Int64((try? await shared.info(forFileAt: at)?.size) ?? 0) : 0

    return Handle(file: file, at: start)
  }

  static func handleClose(_ file: Handle) async {
    try? await file.file.close()
  }

  static func handleRead(_ file: Handle, _ size: Int) async -> String {
    guard size > 0 else { return "" }

    guard
      let chunk = try? await file.file.readChunk(
        fromAbsoluteOffset: file.at,
        length: .bytes(Int64(size))
      )
    else {
      return ""
    }

    file.at += Int64(chunk.readableBytes)

    return String(buffer: chunk)
  }

  static func handleWrite(_ file: Handle, _ data: String) async -> Int {
    let bytes = Array(data.utf8)

    guard
      (try? await file.file.write(contentsOf: bytes, toAbsoluteOffset: file.at))
        != nil
    else {
      return 0
    }

    file.at += Int64(bytes.count)

    return bytes.count
  }

  // `frame` is "start" (absolute), "relative" (from the cursor) or "end" (back from the end)
  static func handleSeek(_ file: Handle, _ offset: Int, _ frame: String) async {
    switch frame {
    case "relative":
      file.at += Int64(offset)
    case "end":
      let size = Int64((try? await file.file.info().size) ?? 0)
      file.at = size - Int64(offset)
    default:
      file.at = Int64(max(0, offset))
    }
  }

  static func handleFlush(_ file: Handle) async {
    // NIOFileSystem has no fsync of its own; the descriptor is the one the syscall wants
    try? await file.file.withUnsafeDescriptor { descriptor in
      _ = fsync(descriptor.rawValue)
    }
  }

  static func handleClear(_ file: Handle, _ size: Int) async {
    try? await file.file.resize(to: .bytes(Int64(max(0, size))))
    file.at = min(file.at, Int64(max(0, size)))
  }

  // ---- streams ----

  static func readerOpen(_ path: String, _ start: Int, _ size: Int) async -> Reader {
    guard let file = try? await shared.openFile(forReadingAt: FilePath(path))
    else {
      fatalError("file reader open: \(path)")
    }

    return Reader(
      file: file,
      at: Int64(max(0, start)),
      left: size > 0 ? Int64(size) : nil
    )
  }

  // the next chunk, or the empty text at the end of the stream (or of the window)
  static func readerNext(_ stream: Reader) async -> String {
    let want: Int64
    if let left = stream.left {
      if left <= 0 { return "" }
      want = min(left, 65536)
    } else {
      want = 65536
    }

    guard
      let chunk = try? await stream.file.readChunk(
        fromAbsoluteOffset: stream.at,
        length: .bytes(want)
      )
    else {
      return ""
    }

    stream.at += Int64(chunk.readableBytes)

    if let left = stream.left {
      stream.left = left - Int64(chunk.readableBytes)
    }

    return String(buffer: chunk)
  }

  static func readerClose(_ stream: Reader) async {
    try? await stream.file.close()
  }

  static func writerOpen(_ path: String, _ append: Bool) async -> Writer {
    let at = FilePath(path)
    let options: OpenOptions.Write =
      append ? .modifyFile(createIfNecessary: true) : .newFile(replaceExisting: true)

    guard let file = try? await shared.openFile(forWritingAt: at, options: options)
    else {
      fatalError("file writer open: \(path)")
    }

    let start: Int64 =
      append ? Int64((try? await shared.info(forFileAt: at)?.size) ?? 0) : 0

    return Writer(file: file, at: start)
  }

  static func writerPush(_ stream: Writer, _ data: String) async {
    let bytes = Array(data.utf8)

    guard
      (try? await stream.file.write(contentsOf: bytes, toAbsoluteOffset: stream.at))
        != nil
    else {
      return
    }

    stream.at += Int64(bytes.count)
  }

  static func writerClose(_ stream: Writer) async {
    try? await stream.file.close()
  }
}

File: deck/seed/code/native/kotlin/file/runtime/aio.kt
// Asynchronous filesystem runtime for the kotlin target: java.nio.file and java.nio.channels, run on
// `Dispatchers.IO` through kotlinx-coroutines.
//
// WHY A DISPATCHER AND NOT A CHANNEL. The JVM's one genuinely asynchronous filesystem type,
// AsynchronousFileChannel, covers plain files and nothing else: no directory walk, no metadata, no symlink, no
// permission, no watch. Half this surface would have to be blocking anyway, and a surface that is asynchronous in
// half its calls is worse than one that is honest. `withContext(Dispatchers.IO)` on a pool that grows for blocked
// threads is what the kotlin ecosystem does for filesystem work, and it is what Ktor itself does.
//
// So `suspend` here means the caller's coroutine is not blocked, which is the promise the Term API makes. It does
// not mean the kernel is doing the waiting, which is what tokio::fs and NIOFileSystem give the other two targets.
// That difference is named in note/term/stdlib/native-async-file-and-server.md rather than papered over.
//
// TOTAL, LIKE THE REST OF THE NATIVE SURFACE: a failure reads as the empty / false / zero answer, except a handle
// open, where a missing file cannot be papered over.
//
// Reached only through the public file API.
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.LinkOption
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.nio.file.attribute.PosixFilePermissions
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object aio {
  // An open file: the channel plus the cursor. A class, not a data holder, because the emitted `file` record is a
  // value and two copies of it must be two views of ONE open file rather than two cursors.
  class Handle(val channel: FileChannel, var at: Long)

  // A read stream: the channel, where it is, and how much of the window is left (null for "to the end").
  class Reader(val channel: FileChannel, var at: Long, var left: Long?)

  class Writer(val channel: FileChannel, var at: Long)

  private fun at(path: String): Path = Paths.get(path)

  // ---- whole file ----

  suspend fun fileRead(path: String): String = withContext(Dispatchers.IO) {
    try {
      Files.readString(at(path))
    } catch (error: Exception) {
      ""
    }
  }

  suspend fun fileReadBytes(path: String): ByteArray = withContext(Dispatchers.IO) {
    try {
      Files.readAllBytes(at(path))
    } catch (error: Exception) {
      ByteArray(0)
    }
  }

  suspend fun fileWrite(path: String, data: String) {
    fileWriteBytes(path, data.toByteArray(Charsets.UTF_8))
  }

  suspend fun fileWriteBytes(path: String, data: ByteArray) {
    withContext(Dispatchers.IO) {
      try {
        Files.write(at(path), data)
      } catch (error: Exception) {
        null
      }
    }
  }

  suspend fun fileAppend(path: String, data: String) {
    withContext(Dispatchers.IO) {
      try {
        Files.write(
          at(path),
          data.toByteArray(Charsets.UTF_8),
          StandardOpenOption.CREATE,
          StandardOpenOption.APPEND,
        )
      } catch (error: Exception) {
        null
      }
    }
  }

  // `deep` copies a directory tree; without it this is the single-file copy node's copyFile means
  suspend fun fileCopy(from: String, to: String, deep: Boolean) {
    withContext(Dispatchers.IO) {
      try {
        if (deep) {
          copyTree(at(from), at(to))
        } else {
          Files.copy(at(from), at(to), StandardCopyOption.REPLACE_EXISTING)
        }
      } catch (error: Exception) {
        null
      }
    }
  }

  private fun copyTree(from: Path, to: Path) {
    if (!Files.isDirectory(from)) {
      Files.copy(from, to, StandardCopyOption.REPLACE_EXISTING)

      return
    }

    Files.createDirectories(to)
    Files.newDirectoryStream(from).use { entries ->
      for (entry in entries) {
        copyTree(entry, to.resolve(entry.fileName))
      }
    }
  }

  suspend fun fileMove(from: String, to: String) {
    withContext(Dispatchers.IO) {
      try {
        Files.move(at(from), at(to), StandardCopyOption.REPLACE_EXISTING)
      } catch (error: Exception) {
        null
      }
    }
  }

  // `deep` removes a directory tree. Without it a directory is still removed when empty, so `remove` means the
  // same thing on a file and on a directory, the way node's `rm` does.
  suspend fun fileRemove(path: String, deep: Boolean) {
    withContext(Dispatchers.IO) {
      try {
        if (deep) {
          removeTree(at(path))
        } else {
          Files.deleteIfExists(at(path))
        }
      } catch (error: Exception) {
        null
      }
    }
  }

  private fun removeTree(path: Path) {
    if (Files.isDirectory(path, LinkOption.NOFOLLOW_LINKS)) {
      Files.newDirectoryStream(path).use { entries ->
        for (entry in entries) {
          removeTree(entry)
        }
      }
    }

    Files.deleteIfExists(path)
  }

  // an empty `kind` asks only whether the path exists; "file", "directory" and "link" ask what it is
  suspend fun fileTest(path: String, kind: String): Boolean = withContext(Dispatchers.IO) {
    try {
      when (kind) {
        "" -> Files.exists(at(path))
        "file" -> Files.isRegularFile(at(path))
        "directory" -> Files.isDirectory(at(path))
        "link" -> Files.isSymbolicLink(at(path))
        else -> Files.exists(at(path))
      }
    } catch (error: Exception) {
      false
    }
  }

  // ---- link ----

  suspend fun linkMake(from: String, to: String, hard: Boolean) {
    withContext(Dispatchers.IO) {
      try {
        if (hard) {
          Files.createLink(at(to), at(from))
        } else {
          Files.createSymbolicLink(at(to), at(from))
        }
      } catch (error: Exception) {
        null
      }
    }
  }

  suspend fun linkRead(path: String): String = withContext(Dispatchers.IO) {
    try {
      Files.readSymbolicLink(at(path)).toString()
    } catch (error: Exception) {
      ""
    }
  }

  // ---- permission and owner ----

  // the raw unix mode, the same number chmod takes, rebuilt from the POSIX permission set the JVM reports
  suspend fun permissionRead(path: String): Long = withContext(Dispatchers.IO) {
    try {
      var mode = 0L

      for (one in Files.getPosixFilePermissions(at(path))) {
        mode = mode or
          when (one) {
            java.nio.file.attribute.PosixFilePermission.OWNER_READ -> 0x100L
            java.nio.file.attribute.PosixFilePermission.OWNER_WRITE -> 0x80L
            java.nio.file.attribute.PosixFilePermission.OWNER_EXECUTE -> 0x40L
            java.nio.file.attribute.PosixFilePermission.GROUP_READ -> 0x20L
            java.nio.file.attribute.PosixFilePermission.GROUP_WRITE -> 0x10L
            java.nio.file.attribute.PosixFilePermission.GROUP_EXECUTE -> 0x8L
            java.nio.file.attribute.PosixFilePermission.OTHERS_READ -> 0x4L
            java.nio.file.attribute.PosixFilePermission.OTHERS_WRITE -> 0x2L
            java.nio.file.attribute.PosixFilePermission.OTHERS_EXECUTE -> 0x1L
          }
      }

      mode
    } catch (error: Exception) {
      0L
    }
  }

  suspend fun permissionWrite(path: String, mode: Long) {
    withContext(Dispatchers.IO) {
      try {
        // the low nine bits as the octal string PosixFilePermissions parses
        val octal = java.lang.Long.toOctalString(mode and 0x1ffL).padStart(3, '0')
        Files.setPosixFilePermissions(
          at(path),
          PosixFilePermissions.fromString(octalToText(octal)),
        )
      } catch (error: Exception) {
        null
      }
    }
  }

  private fun octalToText(octal: String): String {
    val bits = arrayOf("---", "--x", "-w-", "-wx", "r--", "r-x", "rw-", "rwx")
    val out = StringBuilder()

    for (digit in octal) {
      out.append(bits[digit - '0'])
    }

    return out.toString()
  }

  suspend fun ownerUser(path: String): Long = withContext(Dispatchers.IO) {
    try {
      (Files.getAttribute(at(path), "unix:uid") as Number).toLong()
    } catch (error: Exception) {
      0L
    }
  }

  suspend fun ownerGroup(path: String): Long = withContext(Dispatchers.IO) {
    try {
      (Files.getAttribute(at(path), "unix:gid") as Number).toLong()
    } catch (error: Exception) {
      0L
    }
  }

  suspend fun ownerWrite(path: String, user: Long, group: Long) {
    withContext(Dispatchers.IO) {
      try {
        Files.setAttribute(at(path), "unix:uid", user.toInt())
        Files.setAttribute(at(path), "unix:gid", group.toInt())
      } catch (error: Exception) {
        null
      }
    }
  }

  // ---- temporary ----

  suspend fun temporaryMake(kind: String, prefix: String, suffix: String): String =
    withContext(Dispatchers.IO) {
      try {
        if (kind == "directory") {
          Files.createTempDirectory(prefix).toString()
        } else {
          Files.createTempFile(prefix, suffix).toString()
        }
      } catch (error: Exception) {
        ""
      }
    }

  // ---- handle ----

  suspend fun handleOpen(
    path: String,
    read: Boolean,
    write: Boolean,
    create: Boolean,
    append: Boolean,
    clear: Boolean,
  ): Handle = withContext(Dispatchers.IO) {
    val options = mutableSetOf<StandardOpenOption>()

    if (read) options.add(StandardOpenOption.READ)
    if (write || append) options.add(StandardOpenOption.WRITE)
    if (create) options.add(StandardOpenOption.CREATE)
    if (clear && !append) options.add(StandardOpenOption.TRUNCATE_EXISTING)
    if (options.isEmpty()) options.add(StandardOpenOption.READ)

    val channel = FileChannel.open(at(path), options)
    // appending starts at the end, everything else at the start, which is what the open flags mean elsewhere
    Handle(channel, if (append) channel.size() else 0L)
  }

  suspend fun handleClose(file: Handle) {
    withContext(Dispatchers.IO) {
      try {
        file.channel.close()
      } catch (error: Exception) {
        null
      }
    }
  }

  suspend fun handleRead(file: Handle, size: Long): String = withContext(Dispatchers.IO) {
    if (size <= 0L) return@withContext ""

    try {
      val buffer = java.nio.ByteBuffer.allocate(size.toInt())
      val count = file.channel.read(buffer, file.at)

      if (count <= 0) return@withContext ""

      file.at += count.toLong()
      String(buffer.array(), 0, count, Charsets.UTF_8)
    } catch (error: Exception) {
      ""
    }
  }

  suspend fun handleWrite(file: Handle, data: String): Long = withContext(Dispatchers.IO) {
    try {
      val bytes = data.toByteArray(Charsets.UTF_8)
      val count = file.channel.write(java.nio.ByteBuffer.wrap(bytes), file.at)
      file.at += count.toLong()
      count.toLong()
    } catch (error: Exception) {
      0L
    }
  }

  // `frame` is "start" (absolute), "relative" (from the cursor) or "end" (back from the end)
  suspend fun handleSeek(file: Handle, offset: Long, frame: String) {
    withContext(Dispatchers.IO) {
      try {
        file.at =
          when (frame) {
            "relative" -> file.at + offset
            "end" -> file.channel.size() - offset
            else -> maxOf(0L, offset)
          }
      } catch (error: Exception) {
        null
      }
    }
  }

  suspend fun handleFlush(file: Handle) {
    withContext(Dispatchers.IO) {
      try {
        file.channel.force(false)
      } catch (error: Exception) {
        null
      }
    }
  }

  suspend fun handleClear(file: Handle, size: Long) {
    withContext(Dispatchers.IO) {
      try {
        file.channel.truncate(maxOf(0L, size))
        file.at = minOf(file.at, maxOf(0L, size))
      } catch (error: Exception) {
        null
      }
    }
  }

  // ---- streams ----

  suspend fun readerOpen(path: String, start: Long, size: Long): Reader =
    withContext(Dispatchers.IO) {
      val channel = FileChannel.open(at(path), setOf(StandardOpenOption.READ))
      Reader(channel, maxOf(0L, start), if (size > 0L) size else null)
    }

  // the next chunk, or the empty text at the end of the stream (or of the window)
  suspend fun readerNext(stream: Reader): String = withContext(Dispatchers.IO) {
    val left = stream.left

    if (left != null && left <= 0L) return@withContext ""

    val want = if (left != null) minOf(left, 65536L) else 65536L

    try {
      val buffer = java.nio.ByteBuffer.allocate(want.toInt())
      val count = stream.channel.read(buffer, stream.at)

      if (count <= 0) return@withContext ""

      stream.at += count.toLong()

      if (left != null) {
        stream.left = left - count.toLong()
      }

      String(buffer.array(), 0, count, Charsets.UTF_8)
    } catch (error: Exception) {
      ""
    }
  }

  suspend fun readerClose(stream: Reader) {
    withContext(Dispatchers.IO) {
      try {
        stream.channel.close()
      } catch (error: Exception) {
        null
      }
    }
  }

  suspend fun writerOpen(path: String, append: Boolean): Writer = withContext(Dispatchers.IO) {
    val options = mutableSetOf(StandardOpenOption.WRITE, StandardOpenOption.CREATE)

    if (!append) options.add(StandardOpenOption.TRUNCATE_EXISTING)

    val channel = FileChannel.open(at(path), options)
    Writer(channel, if (append) channel.size() else 0L)
  }

  suspend fun writerPush(stream: Writer, data: String) {
    withContext(Dispatchers.IO) {
      try {
        val bytes = data.toByteArray(Charsets.UTF_8)
        val count = stream.channel.write(java.nio.ByteBuffer.wrap(bytes), stream.at)
        stream.at += count.toLong()
      } catch (error: Exception) {
        null
      }
    }
  }

  suspend fun writerClose(stream: Writer) {
    withContext(Dispatchers.IO) {
      try {
        stream.channel.close()
      } catch (error: Exception) {
        null
      }
    }
  }
}
